### PR TITLE
Add positions to events (issue #37)

### PR DESCRIPTION
**Pull request by wzwywx**
This is a draft PR to improve the diagnostics message as indicated in [ issue 37](https://github.com/kaleidawave/ezno/issues/37). 

I've managed to add the position of the return to the `ReturnedTypeDoesNotMatch` error. 

If the solution is acceptable, I can do the same for the other variants.

### CODE REVIEW

**Comment by kaleidawave** [comment]
Hey, was away, so checking back in now.

This looks great!

image

Would be great to add positions to the rest of the `Event` variants. Unlike the return, they don't have any use in diagnostics/errors right now but will come in very useful in the future.

*BTW if the conditional event variant is difficult then you can skip it.*

**Comment by wzwywx** [comment]
Sure thing. I’ll have a look at the other Event variants as well and see what I can do.
29 Oct 2023 at 17:27 by ***@***.***:

**Comment by kaleidawave** [comment]
Awesome, lmk the time frame you are thinking of. Also #71 changed a lot of files, so remember to update/fast-forward the branch before any changes

**Comment by wzwywx** [comment]
Thanks for the heads up on the change.

At the moment, no concrete time frame I'm afraid.

**Comment by wzwywx** [comment]
I'm working on the `ReadsReference` variant right now and the `synthesise_class_fields` uses it. 

https://github.com/kaleidawave/ezno/blob/511e5e2288e13ad251a70c5989de5e422327756e/checker/src/synthesis/expressions.rs#L857-L861

Should the event store the positions of all the fields?

**Comment by kaleidawave** [comment]
Ah interesting. Classes aren't fully implemented at the moment, I got 70% through then got stuck so there might be some unused non fully finished functions around. I would put a `todo!("get event position")` where the position is needed and I can figure out that part when I finish and fix classes.

**Comment by wzwywx** [comment]
Oh, got it. I’ll put a todo there then and move on to the next variant.
2 Nov 2023 at 18:07 by ***@***.***:

File: checker/src/diagnostics.rs [hunk 1]
```diff
@@ -440,12 +441,17 @@ mod defined_errors_and_warnings {
...
 					position,
+					returned_position,
 					expected_return_type,
 					returned_type,
-				} => Diagnostic::Position {
+				} => Diagnostic::PositionWithAdditionLabels {
 					reason: format!(
 						"Function is expected to return {expected_return_type} but returned {returned_type}",
 					),
+					labels: vec![(
```
Comment (kaleidawave) [inline checker/src/diagnostics.rs:451]: Thinking about this, I think `returned_position` should be the main diagnostic position. Could we swap the values for the main `position` and label?

Then the main reason should be "Cannot return {returned type} from function expected to return {expected_return_type}" and the label one should be "function annotated to return {expected_return_type} here"

File: checker/src/events/helpers.rs [hunk 2]
```diff
@@ -55,8 +55,9 @@ pub(crate) fn get_return_from_events<'a, T: crate::ReadFromFS, M: crate::ASTImpl
 										&checking_data.types,
 										false,
 									),
-								position: annotation_span,
-								// TODO event position here #37
+								position: annotation_span.clone(),
```
Comment (kaleidawave) [inline checker/src/events/helpers.rs:58]: Maybe rename `position` to `annotation_position`

File: checker/src/events/helpers.rs [hunk 3]
```diff
@@ -55,8 +55,9 @@ pub(crate) fn get_return_from_events<'a, T: crate::ReadFromFS, M: crate::ASTImpl
...
+								// TODO test with return_position
+								returned_position: position.clone(), // TODO event position here #37
```
Comment (kaleidawave) [inline checker/src/events/helpers.rs:60]: Looking perfect 🤩, can remove TODO the comment now!

**Review by kaleidawave — COMMENTED** [review]
Looking great so far. So just `Event::Setter` left? Left a few minor suggestions around the diagnostic messaging. I have some good ideas for how to use these new getter and setter positions 👀 .

**Comment by wzwywx** [comment]
> Looking great so far. So just `Event::Setter` left? Left a few minor suggestions around the diagnostic messaging. I have some good ideas for how to use these new getter and setter positions 👀 .

I believe `Setter`, `CallsType`, `Throw`, `Conditionally`, and `CreateObject` are left. 

Thanks for the suggestions! I'll look into those first before implementing the position for `Setter`.

File: checker/src/diagnostics.rs (hunk 1 above)
Comment (wzwywx) [reply checker/src/diagnostics.rs:451]: On it, let me swap them and see how the new message looks.

Comment (wzwywx) [reply checker/src/diagnostics.rs:451]: How does this look?

Screenshot 2023-11-04 150607

**Comment by kaleidawave** [comment]
*ignore the test issue, that is something I broke*

Replying to setter position:

> For now, because of the existence of some properties with non-obvious (or perhaps absent) positions,
the position in Event::Setter will be optional i.e. `Option<SpanWithSource>`

Yes that's perfect. The `register_property` function is a bit WIP and has some problems with it. If it is possible put a comment next to any `None`s passed as positions to the setter and I will have a think if positions are even needed in that case.

**Comment by wzwywx** [comment]
Noted. I’ll have a look at what I can find and put a comment where there’s a None. I did put comments before some of the Nones, I think, but definitely not all of them.

8 Nov 2023 at 18:33 by ***@***.***:

**Comment by wzwywx** [comment]
I believe all event variants have positions now. Unfortunately, wiring up the positions for `Setter`, `Conditionally`, and `CreateObject` are difficult for me to implement so at the moment, most of them only have `None`.

**Comment by kaleidawave** [comment]
Awesome just adjusted to use the new return diagnostics message in the tests https://github.com/kaleidawave/ezno/pull/69/commits/071ce967db0eec0b5183db8c1a705df2b8f4501f

Apart from that all the other tests pass, so you haven't broken anything!!! and have just merged!

Thanks for the informative comments and commit messages. They will be useful for using the new positions on events.

**Comment by kaleidawave** [comment]
Awesome. You are also the first to PR to the type checker part of Ezno! If you have any feedback on what is messy, difficult or could be made clearer that would be really helpful for me to help future contributors!

Also might post this change on Twitter, let me know if/how you want a shout out :)

**Comment by wzwywx** [comment]
I have no particular feedback in mind, unfortunately. Though maybe, I’m not sure if it’s just me being new, but sometimes I was unclear whether certain part of the code is (relatively) complete or a work in progress. There are `todo!()` and `TODO` comment to indicate the WIP-ness, but I was oblivious to those so I chalked it up solely to my incompetence when I can’t make things work when in reality, it’s a mix of lack of skill and working on incomplete code.


13 Nov 2023 at 17:42 by ***@***.***:

**Pull request merged**